src/sinks/s3_common/sink.rs

```diff
@@ -1,68 +1,221 @@
+
+        let mut encoded = BytesMut::new();
+        self.encoder
+            .serialize(event, &mut encoded)
+            .expect("writing to memory");
+
+        EncodedEvent {
+            original,
+            json_size,
+            encoded,
+        }
+    }
+}
+
+impl<Svc, RB, Part, Req> S3Sink<Svc, RB, Part>
 where
-    Svc: Service<RB::Request> + Send + 'static,
+    Svc: Service<Req> + Send + 'static,
     Svc::Future: Send + 'static,
     Svc::Response: DriverResponse + Send + 'static,
     Svc::Error: fmt::Debug + Into<crate::Error> + Send,
-    RB: RequestBuilder<(S3PartitionKey, Vec<Event>)> + Send + Sync + 'static,
-    RB::Error: fmt::Display + Send,
-    RB::Request: Finalizable + MetaDescriptive + Send,
+    RB: RequestBuilder<Request = Req> + Send + Sync + 'static,
+    RB::PartitionKey: Send + 'static,
+    Part: Partitioner<Item = EncodedEvent, Key = Option<RB::PartitionKey>> + Send + Unpin + 'static,
+    Part::Key: Eq + std::hash::Hash + Clone + Send + 'static,
+    Req: MetaDescriptive + Finalizable + Send + 'static,
 {
     async fn run_inner(self: Box<Self>, input: BoxStream<'_, Event>) -> Result<(), ()> {
         let partitioner = self.partitioner;
         let settings = self.batcher_settings;
+        let mut event_encoder = EventEncoder::new(self.transformer, self.serializer.clone());
 
+        let payload_builder = Arc::new(PayloadBuilder::new(
+            self.serializer,
+            self.framer,
+            self.request_builder.compression(),
+        ));
         let builder_limit = NonZeroUsize::new(64);
-        let request_builder = self.request_builder;
+        let request_builder = Arc::new(self.request_builder);
 
         input
-            .batched_partitioned(partitioner, settings)
+            .map(|event| event_encoder.encode(event))
+            .batched_partitioned_with_size_fn(partitioner, settings, |e| e.encoded.len())
             .filter_map(|(key, batch)| async move { key.map(move |k| (k, batch)) })
-            .request_builder(builder_limit, request_builder)
+            .concurrent_map(builder_limit, move |(partition_key, events)| {
+                let payload_builder = Arc::clone(&payload_builder);
+                let request_builder = Arc::clone(&request_builder);
+
+                Box::pin(async move {
+                    let event_count = events.len();
+
+                    let mut original_events = Vec::with_capacity(event_count);
+                    let mut encoded_events = Vec::with_capacity(event_count);
+                    let mut events_estimated_json_size =
+                        telemetry().create_request_count_byte_size();
+                    for e in events {
+                        original_events.push(e.original);
+                        encoded_events.push(e.encoded);
+                        events_estimated_json_size = events_estimated_json_size + &e.json_size;
+                    }
```

Comment:
I'm sure there's a way to do this with combinators but it's doubtful that would improve the code here.